onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_api_helper.cc

```diff
@@ -0,0 +1,152 @@
+ * @param target_device_option target devices users want to use.
+ *
+ * @return Status of the operation.
+ *
+ */
+Status GetTargetDevices(const NnApi& nnapi_handle, TargetDeviceOption target_device_option,
+                        InlinedVector<DeviceWrapper>& devices) {
+  // GetTargetDevices is only supported when NNAPI runtime feature level >= ANEURALNETWORKS_FEATURE_LEVEL_3
+  if (GetNNAPIRuntimeFeatureLevel(nnapi_handle) < ANEURALNETWORKS_FEATURE_LEVEL_3)
+    return Status::OK();
+
+  uint32_t num_devices = 0;
+  RETURN_STATUS_ON_ERROR_WITH_NOTE(
+      nnapi_handle.ANeuralNetworks_getDeviceCount(&num_devices), "Getting count of available devices");
+
+  int32_t cpu_index = -1;
+  for (uint32_t i = 0; i < num_devices; i++) {
+    ANeuralNetworksDevice* device = nullptr;
+    const char* device_name = nullptr;
+    int32_t device_type = 0;
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(
+        nnapi_handle.ANeuralNetworks_getDevice(i, &device), "Getting " + std::to_string(i) + "th device");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi_handle.ANeuralNetworksDevice_getName(device, &device_name),
+                                     "Getting " + std::to_string(i) + "th device's name");
+
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi_handle.ANeuralNetworksDevice_getType(device, &device_type),
+                                     "Getting " + std::to_string(i) + "th device's type");
+
+    int64_t curr_device_feature_level = 0;
+    RETURN_STATUS_ON_ERROR_WITH_NOTE(nnapi_handle.ANeuralNetworksDevice_getFeatureLevel(device, &curr_device_feature_level),
+                                     "Getting " + std::to_string(i) + "th device's feature level");
+
+    // https://developer.android.com/ndk/reference/group/neural-networks#aneuralnetworksdevice_gettype
+    bool device_is_cpu = device_type == ANEURALNETWORKS_DEVICE_CPU;
+    if ((target_device_option == TargetDeviceOption::CPU_DISABLED && device_is_cpu) ||
+        (target_device_option == TargetDeviceOption::CPU_ONLY && !device_is_cpu)) {
+      continue;
+    }
+
+    if (device_is_cpu) {
+      cpu_index = static_cast<int32_t>(devices.size());
+    }
+    devices.push_back({device, std::string(device_name), device_type, curr_device_feature_level});
+  }
+
+  // Put CPU device at the end
+  // 1) It's helpful to accelerate NNAPI compiling, just assuming nnapi-reference has the lowest priority
+  // and it internally skips the last device if it has already found one.
+  // 2) We can easily identify if the last device is nnapi-reference without traversing the whole device list.
+  // and exclude nnapi-reference for better performance if possible.
+  // 3) We can easily log the detail of how op was assigned on NNAPI devices which is helpful for debugging.
+  // Related doc and code https://source.android.com/docs/core/interaction/neural-networks#cpu-usage
+  // and https://android.googlesource.com/platform/frameworks/ml/+/master/nn/runtime/ExecutionPlan.cpp#2303
```

Comment:
thanks for adding the links.
is it possible to use a stable link to the code? e.g., one referring to a commit or tag instead of master

Reply:
Yes. Used a Blob instead.